PowerEditor/src/tools/NppShell/Installer.cpp

```diff
@@ -0,0 +1,329 @@
+#include "pch.h"
+#include "Installer.h"
+
+#include "EditWithNppExplorerCommandHandler.h"
+
+#define GUID_STRING_SIZE 40
+
+using namespace winrt::Windows::ApplicationModel;
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Foundation::Collections;
+using namespace winrt::Windows::Management::Deployment;
+
+using namespace NppShell::Helpers;
+using namespace NppShell::Installer;
+
+const wstring SparsePackageName = L"NotepadPlusPlus";
+const int FirstWindows11BuildNumber = 22000;
+
+#ifdef WIN64
+const wstring ShellExtensionKey = L"Software\\Classes\\*\\shell\\ANotepad++64";
+#else
+const wstring ShellExtensionKey = L"Software\\Classes\\*\\shell\\ANotepad++";
+#endif
+
+struct DOREGSTRUCT {
+    HKEY hive;
+    wstring key;
+    wstring name;
+    wstring value;
+};
+
+bool IsWindows11Installation()
+{
+    wstring keyName = L"SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion";
+    wstring valueName = L"CurrentBuildNumber";
+
+    bool result = false;
+
+    HKEY hkey;
+    HRESULT hResult = RegOpenKeyEx(HKEY_LOCAL_MACHINE, keyName.c_str(), 0, KEY_READ, &hkey);
+
+    if (hResult == ERROR_SUCCESS)
+    {
+        DWORD cbData = 0;
+        hResult = RegGetValue(hkey, nullptr, valueName.c_str(), RRF_RT_REG_SZ, nullptr, nullptr, &cbData);
+
+        if (hResult == ERROR_SUCCESS)
+        {
+            vector<BYTE> buffer(cbData + 1);
+
+            cbData = (DWORD)buffer.size();
+            hResult = RegGetValue(hkey, nullptr, valueName.c_str(), RRF_RT_REG_SZ, nullptr, &buffer[0], &cbData);
+
+            wstring buildNumberString(reinterpret_cast<wchar_t*>(&buffer[0]));
+
+            int buildNumber = stoi(buildNumberString);
+
+            result = buildNumber >= FirstWindows11BuildNumber;
+        }
+    }
+
+    RegCloseKey(hkey);
+
+    return result;
+}
+
+wstring GetCLSIDString()
+{
+    auto uuid = __uuidof(NppShell::CommandHandlers::EditWithNppExplorerCommandHandler);
+
+    LPOLESTR guidString = 0;
+    HRESULT result = StringFromCLSID(uuid, &guidString);
```

Comment:
The caller is responsible for freeing the memory allocated for the string by calling the [CoTaskMemFree](https://learn.microsoft.com/en-us/windows/desktop/api/combaseapi/nf-combaseapi-cotaskmemfree) function.